File: 프로그래머스/2/172927. 광물 캐기/광물 캐기.py
from heapq import heappush, heappop

def solution(picks, minerals):
    answer = 0
    q = []
    total = 0
    for i in range(3):
        total += 5 * picks[i]
        
    N = len(minerals)
    for i in range(0, N, 5):
        d, ir, s = 0, 0, 0
        for j in range(5):
            if i + j >= N or i + j >= total:
                break
                
            if minerals[i+j] == "diamond":
                d += 1
            elif minerals[i+j] == "iron":
                ir += 1
            elif minerals[i+j] == "stone":
                s += 1
        heappush(q, (-d, -ir, -s))
    
    tool = 0
    while q:
        d, ir, s = heappop(q)
        
        d *= -1
        ir *= -1
        s *= -1
        
        while tool < 3 and picks[tool] <= 0:
            tool += 1
                
        if tool >= 3:
            break
        
        print(picks, tool, d, ir, s)
            
        if tool == 0 and picks[tool] > 0:
            answer += d + ir + s
        elif tool == 1 and picks[tool] > 0:
            answer += d * 5 + ir + s
        elif tool == 2 and picks[tool] > 0:
            answer += d * 25 + ir * 5 + s
            
        picks[tool] -= 1
        
    return answer

'''
(5, 0, 0) , (0, 5, 0), (1, 0, 0)
(3, 2, 0), (1, 1, 1)

'''

File: 프로그래머스/2/172927. 광물 캐기/광물 캐기.java
import java.util.PriorityQueue;
import java.util.Comparator;

class Solution {
    static class Mineral{
        int d, ir, s;

        public Mineral(int d, int ir, int s) {
            this.d = d;
            this.ir = ir;
            this.s = s;
        }
    }

    public int solution(int[] picks, String[] minerals) {
        int answer = 0;
        PriorityQueue<Mineral> pq = new PriorityQueue<>(
            (a, b) -> {
                if (a.d != b.d) {
                    return b.d - a.d;
                }
                
                if (a.ir != b.ir) {
                    return b.ir - a.ir;
                }
                return b.s - a.s;
            }
        );
        
        int total = 0;
        for (int i = 0; i < 3; i++) {
            total += 5 * picks[i];
        }
        int N = minerals.length;
        for (int i = 0; i < N; i += 5) {
            int d = 0;
            int ir = 0;
            int s = 0;
            
            for (int j = 0; j < 5; j++) {
                if (i + j >= N || i + j >= total) {
                    break;
                }
                
                if (minerals[i+j].equals("diamond")) {
                    d += 1;
                } else if (minerals[i+j].equals("iron")) {
                    ir += 1;
                } else {
                    s += 1;
                }
            }
            pq.offer(new Mineral(d, ir, s));
            
        }
        
        int tool = 0;
        while (!pq.isEmpty()) {
            Mineral tmp = pq.poll();
            
            while (tool < 3 && picks[tool] <= 0) {
                tool++;
            }
            
            if (tool >=3) {
                break;
            }
            if (tool == 0) {
                answer += tmp.d + tmp.ir + tmp.s;
            } else if (tool == 1) {
                answer += tmp.d * 5 + tmp.ir + tmp.s;
            } else if (tool == 2) {
                answer += tmp.d * 25 + tmp.ir * 5 + tmp.s;
            }
            
            picks[tool]--;
        }
        
        return answer;
    }
    
}
